streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -0,0 +1,184 @@
+        }
+        log.debug("Took {} ms to validate that partition {} exists", time.milliseconds() - start, topicPartition);
+    }
+
+    @Override
+    public void register(final StateRestorer restorer) {
+        stateRestorers.put(restorer.partition(), restorer);
+    }
+
+    public void restore() {
+        final long start = time.milliseconds();
+        try {
+            if (!consumer.subscription().isEmpty()) {
+                throw new IllegalStateException(String.format("Restore consumer should have not subscribed to any partitions (%s) beforehand", consumer.subscription()));
+            }
+            final Map<TopicPartition, Long> endOffsets = consumer.endOffsets(stateRestorers.keySet());
+            consumer.assign(stateRestorers.keySet());
+
+            for (final StateRestorer restorer : stateRestorers.values()) {
+                if (restorer.checkpoint() != null) {
+                    consumer.seek(restorer.partition(), restorer.checkpoint());
+                } else {
+                    consumer.seekToBeginning(Collections.singletonList(restorer.partition()));
+                }
+            }
+
+            final Set<TopicPartition> partitions = new HashSet<>(stateRestorers.keySet());
+            while (!partitions.isEmpty()) {
+                final ConsumerRecords<byte[], byte[]> allRecords = consumer.poll(10);
+                final Iterator<TopicPartition> partitionIterator = partitions.iterator();
+                while (partitionIterator.hasNext()) {
+                    restorePartition(endOffsets, allRecords, partitionIterator);
+                }
+            }
+        } finally {
+            consumer.assign(Collections.<TopicPartition>emptyList());
+            log.debug("Took {} ms to restore active state", time.milliseconds() - start);
+        }
+    }
+
+    @Override
+    public Map<TopicPartition, Long> restoredOffsets() {
+        final Map<TopicPartition, Long> restoredOffsets = new HashMap<>();
+        for (final Map.Entry<TopicPartition, StateRestorer> entry : stateRestorers.entrySet()) {
+            final StateRestorer restorer = entry.getValue();
+            if (restorer.isPersistent()) {
+                restoredOffsets.put(entry.getKey(), restorer.restoredOffset());
+            }
+        }
+        return restoredOffsets;
+    }
+
+    private void restorePartition(final Map<TopicPartition, Long> endOffsets,
+                                  final ConsumerRecords<byte[], byte[]> allRecords,
+                                  final Iterator<TopicPartition> partitionIterator) {
+        final TopicPartition topicPartition = partitionIterator.next();
+        final StateRestorer restorer = stateRestorers.get(topicPartition);
+        final Long endOffset = endOffsets.get(topicPartition);
+        final long pos = processNext(allRecords.records(topicPartition), restorer, endOffset);
+        if (restorer.hasCompleted(pos + 1, endOffset)) {
```

Comment:
Hmm, maybe I read the code wrong, but in the restored offset we are actually storing the fetched offset + 1 so that this offset can be used directly as the starting offset in the future fetch requests right?

```
restoredOffsets.put(storePartition, lastOffset + 1);
```

So if 1) there is no data at all in the topic, then end offset should be -1 and `consumer.position` should be 1; 2) if there are some data but they have all been fetched and offsets committed in the previous run (say log end offset is 10, so the committed offset would be 11), then end offset should be 10 while `consumer.position` should be 11.